test/parallel/test-tls-alert-handling.js

```diff
@@ -78,8 +81,10 @@ function sendBADTLSRecord() {
     socket: socket,
     rejectUnauthorized: false
   }, common.mustCall(function() {
-    socket.write(BAD_RECORD);
-    socket.end();
+    client.write('x');
+    client.on('data', (data) => {
+      socket.end(BAD_RECORD);
+    });
```

Comment:
I think this is the only change that I don’t understand – why is `.write('x')` necessary? Or is it not necessary and does it just expand the test?

Reply:
It's the delay of the BAD_RECORD until `.on(data)` that is necessary. What's happening on the server side at the point that the client handshake completes depends on the protocol version, so the behaviour with a bad record varies, and the relative ordering of the client `secureConnect` and server `secureConnection` are undefined. IIRC, on 1.3 the bad record interrupts the server handshake, causing a `tlsClientError` (thus the `common.mustNotCall()` I introduced for that event) instead of a `secureConnection` event. So, I introduced a data exchange. The server is doing `client.pipe(client)` to echo data back, so I write from the client, and when the data echoes back, I know that the handshake must have completed on the server.